win32/tests/job_object_helper_int/job_object_helper_int.c

```diff
@@ -0,0 +1,179 @@
+// Copyright(C) Microsoft Corporation.All rights reserved.
+
+
+#include <stdlib.h>
+
+#include "windows.h"
+
+#include "macro_utils/macro_utils.h"
+
+#include "testrunnerswitcher.h"
+
+#include "c_pal/timer.h"
+#include "c_pal/job_object_helper.h"
+
+#define MEGABYTE ((size_t)1024 * 1024)
+#define GIGABYTE (MEGABYTE * 1024)
+#define MAX_MEMORY_POSSIBLE (GIGABYTE * 256)
+#define TEST_BUFFER_SIZE (MEGABYTE * 512)
+#define MAX_BUFFERS_BEFORE_FAILURE (MAX_MEMORY_POSSIBLE / TEST_BUFFER_SIZE)
+
+const size_t max_buffers_before_failure = MAX_BUFFERS_BEFORE_FAILURE;
+
+static size_t get_malloc_limit_before_failure()
+{
+    /* malloc until failure and return the count of (512MB) buffers we were able to malloc. */
+    void** buffers = malloc(sizeof(void*) * max_buffers_before_failure);
+    ASSERT_IS_NOT_NULL(buffers);
+
+    size_t max_buffers = 0;
+    for (size_t i = 0; i < max_buffers_before_failure; i++)
+    {
+        buffers[i] = malloc(TEST_BUFFER_SIZE);
+        if (buffers[i] == 0)
+        {
+            max_buffers = i;
+            break;
+        }
+    }
+    ASSERT_ARE_NOT_EQUAL(size_t, 0, max_buffers);
+    for (size_t i = 0; i < max_buffers; i++)
+    {
+        free(buffers[i]);
+    }
+    free(buffers);
+
+    return max_buffers;
+}
+
+
+/* We need something to keep the CPU busy. */
+/* If you list the numbers from 0 to 2^28-1 in binary, you'll find that 3,758,096,384 bits are set. */
+/* Lets verify this. It takes about 5 seconds on my desktop PC. */
+static size_t limit = 1 << 28;
+static size_t expected_bits = 3758096384;
+
+static size_t get_elapsed_milliseconds_for_cpu_bound_task()
+{
+    double start = timer_global_get_elapsed_ms();
+
+    size_t total_bits = 0;
+
+    for (size_t i = 0; i < limit; i++)
+    {
+        size_t this = i;
+        while (this != 0)
+        {
+            total_bits += 1;
+            this = this & (this - 1);
+        }
+    }
+
+    double end = timer_global_get_elapsed_ms();
+    ASSERT_ARE_EQUAL(size_t, expected_bits, total_bits);
+    return (size_t)(end - start);
+}
+
+BEGIN_TEST_SUITE(TEST_SUITE_NAME_FROM_CMAKE)
+
+TEST_SUITE_INITIALIZE(suite_init)
+{
+}
+
+TEST_SUITE_CLEANUP(suite_cleanup)
+{
+}
+
+TEST_FUNCTION_INITIALIZE(init)
+{
+}
+
+TEST_FUNCTION_CLEANUP(cleanup)
+{
+}
+
+TEST_FUNCTION(test_job_object_helper_limit_memory)
+{
+    LogInfo("Each block is %zu MB", TEST_BUFFER_SIZE / MEGABYTE);
+
+    THANDLE(JOB_OBJECT_HELPER) job_object_helper = job_object_helper_create();
+    ASSERT_IS_NOT_NULL(job_object_helper);
+
+    /* Constrain to 100% of physical memory and re-measure. */
```

Comment:
>100 [](http://example.com/codeflow?start=20&length=3)

Hmmm... might I humbly suggest not eating ALL the memory on the machine for this test?

How would this even work on a machine without virtual memory (which our test machines might be, and for sure our production machines are).

step 1: let's eat all the memory until we fail to allocate
step2: ???
step3: profit (other processes that so happen to try to allocate at this time of high memory pressure are unhappy).

So to keep everybody happy... I would suggest setting a limit lower than 100%

Reply:
To be fair, this doesn't allocate _all_ of the physical memory. It allocates as many contiguous 512MB blocks as it can, so it could leave up to 512MB-4KB free. Plus heap fragmentation (which may or may not matter because my MMU knowledge is out of date). 

But, you make a good point.  Let's not blow anything up if we don't have to.